File: ArmLibrary/ThrowData.cs
using System;
using System.IO;
using System.Web.Script.Serialization;
using System.Windows;
using Newtonsoft.Json;

namespace ArmLibrary
{
    public class ThrowData
    {
        #region Properties

        public DataStruct Data;
        public struct DataStruct
        {
            public int TrialNumber { get; set; } //read by Python Script; it tells the gILC to start new (assign zeros to estimation)
            public int SamplingFrequency { get; set; }
            public DateTime DateCalculated { get; set; }
            public DateTime DateExecuted { get; set; }
            public int ArraySize { get; set; }
            public JointStruct Shoulder;
            public JointStruct Elbow;
            
        }
        public struct JointStruct
        {
            public float[] Cmd { get; set; }
            public float[] Est { get; set; }
            public float[] Ref { get; set; }
            public int[] Sensor { get; set; }
            public float[] Time { get; set; }
        }

        #endregion

        #region Constructor
        public ThrowData()
        {
            //Data.Shoulder.Sensor = new int[] { 1, 13, 33, 77, 3, 13, 19 };
            //Data.Elbow.Sensor = new int[] { 77, 77, 77, 77, 77, 77, 77 };
            //Data.DateCalculated = DateTime.Now;
            //Data.DateExecuted = Data.DateCalculated;
        }

        #endregion

        #region Methods

        /// <summary>
            /// Sets the trial number to zero, sets the sampling frequency, and Initializes the Json file to be used by Python for the first throw
            /// </summary>
            /// <param name="path"></param>
        public void WriteFirstThrowDataToJson(string path, float[] time)
        {
            Data.TrialNumber = 0;
            Data.ArraySize = GlobalVariables.ARRAY_SIZE;
            Data.SamplingFrequency = GlobalVariables.SAMPLING_FREQUENCY;
            Data.Shoulder.Time = time;
            Data.Elbow.Time = time;
            WriteDataToJsonFile(path);
        }

        /// <summary>
        /// Writes the throw data to the Json file to be used by Python for the next throw
        /// </summary>
        /// <param name="path"></param>
        public void WriteDataToJsonFile(string path)
        {
            var settings = new JsonSerializerSettings { DateFormatString = "yyyy-MM-ddTHH:mm:ss.ffffffZ" };
            var jsonstring = JsonConvert.SerializeObject(Data, settings);
            File.WriteAllText(path, jsonstring);
        }

        /// <summary>
        /// Reads Json File and stores throw Data
        /// </summary>
        /// <param name="path"></param>
        public void ReadJsonFile(string path) //filepath should be of format: @"C:\movie.json"
        {
            string json = File.ReadAllText(path);
            var json_serializer = new JavaScriptSerializer();
            Data = json_serializer.Deserialize<DataStruct>(json);
        }
        

        #endregion
    }
}

File: Python/ThrowData.py
import json
import datetime


class JointStruct:
    def __init__(self):
        self.Cmd = []       # u(k, i)
        self.Est = []       # a(k, i)
        self.Ref = []       # r(i)
        self.Sensor = []    # y(k, i)
        self.Time = []      # t(i)
        self.json = {}

    def updatedatafromjson(self, jsondata):
        self.Cmd = jsondata["Cmd"]
        self.Est = jsondata["Est"]
        self.Ref = jsondata["Ref"]
        self.Sensor = jsondata["Sensor"]
        self.Time = jsondata["Time"]
        self.json = jsondata

    def refreshjsonfromdata(self):
        self.json["Cmd"] = self.Cmd
        self.json["Est"] = self.Est
        self.json["Ref"] = self.Ref
        self.json["Sensor"] = self.Sensor
        self.json["Time"] = self.Time
        return self.json


class ThrowData:
    def __init__(self):
        self.TrialNumber = -1
        self.SamplingFrequency = 0
        self.ArraySize = 0
        self.DateCalculated = {}
        self.DateExecuted = {}
        self.Shoulder = JointStruct()
        self.Elbow = JointStruct()
        self.json = {}
        self.json["TrialNumber"] = 0
        self.json["Shoulder"] = {}
        self.json["Elbow"] = {}

    def updatedatafromjson(self, jsondata):
        self.TrialNumber = jsondata["TrialNumber"]
        self.SamplingFrequency = jsondata["SamplingFrequency"]
        self.ArraySize = jsondata["ArraySize"]
        self.DateCalculated = datetime.datetime.strptime(jsondata["DateCalculated"], '%Y-%m-%dT%H:%M:%S.%fZ')
        self.DateExecuted = datetime.datetime.strptime(jsondata["DateExecuted"], '%Y-%m-%dT%H:%M:%S.%fZ')
        self.Shoulder.updatedatafromjson(jsondata["Shoulder"])
        self.Elbow.updatedatafromjson(jsondata["Elbow"])
        self.json = jsondata

    def readfile(self, filename):
        jsonfile = open(filename, "r")
        jsondata = json.load(jsonfile)
        self.updatedatafromjson(jsondata)
        jsonfile.close()

    def writefile(self, filename):
        jsonfile = open(filename, "w")
        self.refreshjsonfromdata()
        json.dump(self.json, jsonfile)
        jsonfile.close()

    def refreshjsonfromdata(self):   # when adding elements, the string is added to beginning, not the end
        self.json["Elbow"] = self.Elbow.refreshjsonfromdata()
        self.json["Shoulder"] = self.Shoulder.refreshjsonfromdata()
        self.json["TrialNumber"] = self.TrialNumber
        self.json["SamplingFrequency"] = self.SamplingFrequency
        self.json["ArraySize"] = self.ArraySize
        self.json["DateCalculated"] = self.DateCalculated.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
        self.json["DateExecuted"] = self.DateExecuted.strftime('%Y-%m-%dT%H:%M:%S.%fZ')


